compiler/lib/ref_unboxing.ml

```diff
@@ -0,0 +1,191 @@
+open! Stdlib
+open Code
+
+(*
+ocamlc does not perform reference unboxing when emitting debugging
+information. Inlining can also enable additional reference unboxing.
+
+We currently does not unbox references which are used within the scope
+of an exception handler. This should often not result in significant
+performance improvements, and is tricky to get right. Indeed, we would
+need to introduce variables for these references right before the
+[Pushtrap], and then add [Assign] instructions to keep their contents
+up to date whenever a reference is updated.
+*)
+
+let debug = Debug.find "unbox-refs"
+
+let times = Debug.find "times"
+
+let stats = Debug.find "stats"
+
+let rewrite refs block m =
+  let m, l =
+    List.fold_left
+      ~f:(fun (m, rem) i ->
+        match i with
+        | Let (x, Block (0, [| y |], (NotArray | Unknown), Maybe_mutable))
+          when Var.Set.mem x refs -> Var.Map.add x y m, rem
+        | Let (y, Field (x, 0, Non_float)) when Var.Map.mem x m ->
+            (* Optimized away by Phisimpl *)
+            m, Let (y, Prim (Extern "%identity", [ Pv (Var.Map.find x m) ])) :: rem
+        | Offset_ref (x, n) when Var.Map.mem x m ->
+            let y = Var.fresh () in
+            ( Var.Map.add x y m
+            , Let
+                ( y
+                , Prim
+                    ( Extern "%int_add"
+                    , [ Pv (Var.Map.find x m); Pc (Int (Targetint.of_int_exn n)) ] ) )
+              :: rem )
+        | Set_field (x, 0, Non_float, y) when Var.Map.mem x m -> Var.Map.add x y m, rem
+        | Event _
+          when match rem with
+               | Event _ :: _ -> true
+               | _ -> false -> m, rem
+        | _ -> m, i :: rem)
+      block.body
+      ~init:(m, [])
+  in
+  m, List.rev l
+
+let rewrite_cont relevant_vars vars (pc', args) =
+  let refs, _ = Int.Hashtbl.find relevant_vars pc' in
+  let vars = Var.Map.filter (fun x _ -> Var.Set.mem x refs) vars in
+  pc', List.map ~f:snd (Var.Map.bindings vars) @ args
+
+let rewrite_function p variables pc =
+  let relevant_vars = Int.Hashtbl.create 16 in
+  let g = Structure.(dominator_tree (build_graph p.blocks pc)) in
+  let rec traverse_tree g pc vars =
+    let block = Addr.Map.find pc p.blocks in
+    let vars' =
+      List.fold_left
+        ~f:(fun s i ->
+          match i with
+          | Let (x, Block (0, [| _ |], (NotArray | Unknown), Maybe_mutable))
+            when Var.Hashtbl.mem variables x -> Var.Set.add x s
+          | _ -> s)
+        ~init:vars
+        block.body
+    in
+    Int.Hashtbl.add relevant_vars pc (vars, vars');
+    Addr.Set.iter (fun pc' -> traverse_tree g pc' vars') (Structure.get_edges g pc)
+  in
+  traverse_tree g pc Var.Set.empty;
+  let rec traverse_tree' g pc blocks =
+    let block = Addr.Map.find pc p.blocks in
+    let vars, refs = Int.Hashtbl.find relevant_vars pc in
```

Comment:
We have line 54 
```ocaml
let refs, _ = Int.Hashtbl.find relevant_vars pc' in
```
and line 78
```ocaml
 let vars, refs = Int.Hashtbl.find relevant_vars pc in
```

Using `refs` in both position when matching the tuple is a bit confusing. I think changing variable names with something more explicit would help.